causal_testing/json_front/json_class.py

```diff
@@ -270,6 +270,9 @@ def _execute_test_case(
         failed = False
 
         estimation_model = self._setup_test(causal_test_case=causal_test_case, test=test)
+        if "formula" in test:
+            if not estimation_model.validate_formula(self.causal_specification.causal_dag):
+                raise ValueError("Formula covariates do not satisfy the constructive back-door criterion.")
```

Comment:
Given that this is what this PR is meant to do, I think it'd be good to have this line covered by a test to make sure that invalid adjustment sets really are picked up. If you have a DAG "X -> Y; Z -> X; Z -> Y" and a causal test with formula "Y ~ X", that should set it off.